qlib/backtest/order.py

```diff
@@ -0,0 +1,357 @@
+        Case 2:
+        1. The outer strategy's decision is available at the start of the interval
+        2. Same as `case 1.3`
+    """
+
+    def __init__(self, strategy: BaseStrategy):
+        """
+        Parameters
+        ----------
+        strategy : BaseStrategy
+            The strategy who make the decision
+        """
+        self.strategy = strategy
+
+    def get_decision(self) -> List[object]:
+        """
+        get the **concrete decision**  (e.g. execution orders)
+        This will be called by the inner strategy
+
+        Returns
+        -------
+        List[object]:
+            The decision result. Typically it is some orders
+            Example:
+                []:
+                    Decision not available
+                [concrete_decision]:
+                    available
+        """
+        raise NotImplementedError(f"This type of input is not supported")
+
+    def update(self, trade_calendar: TradeCalendarManager) -> Union["BaseTradeDecision", None]:
+        """
+        Be called at the **start** of each step
+
+        Parameters
+        ----------
+        trade_calendar : TradeCalendarManager
+            The calendar of the **inner strategy**!!!!!
+
+        Returns
+        -------
+        None:
+            No update, use previous decision(or unavailable)
+        BaseTradeDecision:
+            New update, use new decision
+        """
+        return self.strategy.update_trade_decision(self, trade_calendar)
+
+    def get_range_limit(self) -> Tuple[int, int]:
+        """
+        return the expected step range for limiting the decision execution time
+        Both left and right are **closed**
+
+        Returns
+        -------
+        Tuple[int, int]:
+
+        Raises
+        ------
+        NotImplementedError:
+            If the decision can't provide a unified start and end
+        """
+        raise NotImplementedError(f"Please implement the `func` method")
+
+    def empty(self) -> bool:
+        return len(self.get_decision()) == 0
+
+
+class TradeDecisionWO(BaseTradeDecision):
+    """
+    Trade Decision (W)ith (O)rder.
+    Besides, the time_range is also included.
+    """
+
+    def __init__(self, order_list: List[Order], strategy: BaseStrategy, idx_range: Tuple = None):
+        super().__init__(strategy)
+        self.order_list = order_list
+        self.idx_range = idx_range
+
+    def get_range_limit(self) -> Tuple[int, int]:
+        if self.idx_range is None:
+            # Default to get full index
+            raise NotImplementedError(f"The decision didn't provide an index range")
+        return self.idx_range
+
+    def get_decision(self) -> List[object]:
+        return self.order_list
+
+    def __repr__(self) -> str:
+        return f"strategy: {self.strategy}; idx_range: {self.idx_range}; order_list[{len(self.order_list)}]"
+
+
+# TODO: the orders below need to be discussed ------------------------------------
```

Comment:
If we don't have time to change it into the right format. 
We could leave it in the historical commit and mark the commit number here

Reply:
# NOTE:  commit:XXXX XXXX